### ISSUE TITLE
Error 500 when submitting order: Unknown offset 0 for collection Order

### ISSUE THREAD

**Issue by LukasK13** [mollie/PrestaShop#266]
### Expected behavior
Payment executing

### Actual behavior
When submitting the order I get error 500. Debug mode brings up the following message:
<img width="992" alt="Bildschirmfoto 2020-12-18 um 18 13 00" src="https://user-images.githubusercontent.com/24586740/102649762-12a3de00-416a-11eb-8e5c-512059da35d4.png">

### Steps to reproduce the behavior
I am new to mollie and just started to set up the payment method. I have an verified mollie account and pasted the live API keys into prestashop. I enabled both Apple Pay and credit card payment. As soon as a customer submits an order with either Apple Pay or credit card, the error is thrown. I'm not sure if this is a problem with 4.1.3 (tested also with 4.1.2) or if I misconfigured anything. Thanks for your help.

### Environment
* PrestaShop version: 1.7.7.0
* Mollie module version: 4.1.3
* PHP version: 7.3

**Comment by Saimis777** [comment]
Hello @LukasK13 thank you for your message.

And could you clarify if this issue occurs only with Apple Pay and credit card? Have you tried enabling other methods? Do they work or it also throws the same error? Also, are you running the default theme, or some custom one? And one more question, do you have any other payment modules enabled at the moment, or just Mollie? If you have any other payment modules enabled, have you tried disabling the other modules, just leaving Mollie working? Sorry for the extensive questions.

Will be looking forward to your reply!

Best regards,
Invertus Support team

**Comment by LukasK13** [comment]
Thank you for your ideas and you help.

To your questions:
1. The same issue occurs, when I enable bank transfer. So the issue is not limited to the payment method but occurs generally for all mollie payments.
2. I'm using the default theme with some CSS modifications I did by myself. In my opinion, these modifications should not impact the payment.
3. I have enabled the PayPal module and bank transfer module. Disabling both modules unfortunately does not fix the issue.

Do you have any further suggestions?

**Comment by Saimis777** [comment]
Thank you for that, and in the payment preferences, are all the countries are enabled? (Adding screenshot) Or some of them are disabled?
![image](https://user-images.githubusercontent.com/75423142/102761651-d9dd5200-437f-11eb-8758-1bddcb8c89d5.png)

Have you tried re-installing Mollie itself? You can download the latest version of it here - https://github.com/mollie/Prestashop/releases

Also, just to replicate the issue, I should proceed with adding a product to my cart, and proceed with making the payment by selecting any payment method from Mollie?

Looking forward to your reply!

Best regards,
Invertus Support team

**Comment by LukasK13** [comment]
By default, only germany (my country) was enabled. However, enabling all other countries did not fix the issue.
I reinstalled the latest version of mollie. Issue still persists.
If I understand the debug message correctly, there is an issue reading the shipping information. Still, I have no idea how to look into this.
Thanks for your help.

**Comment by Saimis777** [comment]
Thank you.

Would you be able to write us at - support@invertus.eu with your Front office/Back office, and FTP details, so we could investigate the issue further? As from what I see at the moment everything should be working as it is, but without actually checking out the shop, and testing it locally we can't really give a solution to this.

Best regards,
Invertus Support team

**Comment by LukasK13** [comment]
Sorry for my late reply.
I'm running prestashop on my private server in a docker container besides several other private docker containers. For this reason I would prefer not to give any credentials away. I'm aware that this limits the support I can get from you.

Today, I had a deeper look into the debug output of prestashop that I posted above. The error occurs [here](https://github.com/mollie/PrestaShop/blob/a9dc0fb6af12633d58c54ad50a07c0c4b523a055/src/Service/ShipmentService.php#L88) when mollie is trying to retrieve information on the order. This call leads the creation of a [new PrestashopCollection](https://github.com/PrestaShop/PrestaShop/blob/32c6b329c51ba07a3d2b1ba938f0632375ae3234/classes/order/Order.php#L1583) for all orders which is filtered for the given order reference id. The array indexing [here](https://github.com/mollie/PrestaShop/blob/a9dc0fb6af12633d58c54ad50a07c0c4b523a055/src/Service/ShipmentService.php#L88) leads to the execution of [this](https://github.com/PrestaShop/PrestaShop/blob/32c6b329c51ba07a3d2b1ba938f0632375ae3234/classes/PrestaShopCollection.php#L529) method of the previously PrestashopCollection.
However, the database select query does not return a result which is the reason why an error is thrown.

A further investigation of the database query logs reveals that the new order is added to the database after the order retrieval by mollie. So mollie queries the order details from the database before they have been added to the database.

You can see the prestashop debug output below. Additionally, I tweaked the PrestaShopCollection class to print the database query which is shown in the first line of the screenshot.
<img width="992" alt="Bildschirmfoto 2020-12-22 um 19 33 42" src="https://user-images.githubusercontent.com/24586740/102921481-bbb14800-448c-11eb-8cb0-ea6ee91b0d64.png">

The relevant part of the DB query log:
```
2020-12-22T18:09:20.851144Z        52 Query     SELECT * FROM `orders` a0 WHERE (a0.`reference` = 'IOCGFUUKI')
2020-12-22T18:09:20.859674Z	   52 Query	INSERT INTO `orders` (`id_address_delivery`, `id_address_invoice`, `id_cart`, `id_currency`, `id_shop_group`, `id_shop`, `id_lang`, `id_customer`, `id_carrier`, `current_state`, `secure_key`, `payment`, `module`, `recyclable`, `gift`, `gift_message`, `mobile_theme`, `total_discounts`, `total_discounts_tax_incl`, `total_discounts_tax_excl`, `total_paid`, `total_paid_tax_incl`, `total_paid_tax_excl`, `total_paid_real`, `total_products`, `total_products_wt`, `total_shipping`, `total_shipping_tax_incl`, `total_shipping_tax_excl`, `carrier_tax_rate`, `total_wrapping`, `total_wrapping_tax_incl`, `total_wrapping_tax_excl`, `round_mode`, `round_type`, `shipping_number`, `conversion_rate`, `invoice_number`, `delivery_number`, `invoice_date`, `delivery_date`, `valid`, `reference`, `date_add`, `date_upd`) VALUES ('14', '14', '64', '1', '1', '1', '1', '19', '9', '0', '13f039bed27ffa172e5d51255c0bb50e', 'Apple Pay', 'mollie', '0', '0', NULL, '0', '0', '0', '0', '20.15', '20.15', '17.91', '0', '11.76', '14', '6.15', '6.15', '6.15', '0', '0', '0', '0', '2', '1', NULL, '1', '0', '0', '0000-00-00 00:00:00', '0000-00-00 00:00:00', '0', 'IOCGFUUKI', '2020-12-22 19:09:20', '2020-12-22 19:09:20')
```

I hope this helps you to look further into this issue

**Comment by LukasK13** [comment]
Further debugging provided me the following insights:

As shown in the stack trace in my initial bug report the error occurs due to the execution of the hook `ActionOrderStatusUpdate`. This hook is called for a status update of the submitted order to the state `Awaiting Mollie payment`. The hook is able to handle missing shipping information as it becomes obvious in [this](https://github.com/mollie/PrestaShop/blob/a9dc0fb6af12633d58c54ad50a07c0c4b523a055/mollie.php#L1152) line. This led me to trying the following:

I replaced [this line](https://github.com/mollie/PrestaShop/blob/a9dc0fb6af12633d58c54ad50a07c0c4b523a055/src/Service/ShipmentService.php#L88) by
```
return null;
```

Doing so actually **fixed the bug** and the mollie checkout page appeared (Note: I aborted the payment at this point and did not complete the order).

So to summarise, I think [this line](https://github.com/mollie/PrestaShop/blob/a9dc0fb6af12633d58c54ad50a07c0c4b523a055/src/Service/ShipmentService.php#L88) should be modified. In detail, the result should be checked before array indexing happens.

As I started using mollie with prestashop v1.7.7, I'm not sure if this issue is due to changes in prestashop.

**Comment by LukasK13** [comment]
The method `getFirst()` of PrestaShopCollection allows a failproof retrieval of the first element of the result array. Using this method would fix the issue. PR following...

**Comment by Saimis777** [comment]
@LukasK13 We are really thankful for your investigation and insights, I will pass this to our development team to implement this bug fix!

Will update this thread once we implement your given information.

Thank you once again.

--
Best regards,
Invertus Support team

**Comment by Saimis777** [comment]
@LukasK13 Also, you can implement these changes to fix the issue:

$orders = Order::getByReference($orderReference);
/** @var Order $order */
$order = $orders->getFirst();
if (!Validate::isLoadedObject($order)) {
	return null;
}


As I can see this was already implemented in the upcoming release, but before that, if there is a need, this can be added in order to fix the issue before the upcoming release is published.

--
Best regards,
Invertus Support team

**Comment by LukasK13** [comment]
Thanks for confirming.
Do you have an idea why I'm currently the only user having this issue?
In my understanding, every user of mollie should experience this issue.

**Comment by Saimis777** [comment]
Sorry, can't really confirm at the moment on why this has happened only with your installation, sorry for the caused inconveniences, the issue will be fixed in the next release, please let us know if you will encounter any issues, or notice any bugs, any feedback is greatly appreciated!

--
Best regards,
Invertus Support team

**Comment by Saimis777** [comment]
I will close this issue for now, if you will feel like having any more questions, or will need any other assistance, please let us know!

--
Best regards,
Invertus Support team

**Comment by Saimis777** [comment]
Hello @LukasK13, hope you're having a nice day!

The official Mollie 4.2.0 was just released, the fix was implemented, you can download the latest version here - https://github.com/mollie/PrestaShop/releases/tag/4.2.0

Any feedback regarding the updated version is greatly appreciated!

--
Best regards,
Invertus Support team
